developer_docs/how_to_create_a_plugin.asciidoc

```diff
@@ -0,0 +1,2124 @@
+foreman plugin that adds hooks.
+
+[[using-the-example-plugin]]
+== Using the example plugin
+
+
+There is a fully working example plugin which you can clone to get you
+started. It contains examples of many of the types of behaviour that you
+might want to do in a plugin, such as adding new models, overriding
+views, extending controllers, adding permissions and menu items, and so
+on. The README contains a list of it's current behaviour. To get started
+building your first Foreman Plugin run the following command:
+
+....
+git clone https://github.com/theforeman/foreman_plugin_template my_plugin
+....
+
+A new directory my_plugin is created for the plugin. Now go into this
+directory and use the rename script to change all references to
+ForemanPluginTemplate to MyPlugin:
+
+....
+cd my_plugin; ./rename.rb my_plugin
+....
+
+[[installing-the-plugin]]
+== Installing the plugin
+
+
+It's best to test a plugin on a development installation of Foreman, as
+it loads code on the fly and doesn't require building and installing
+your plugin as a gem. http://theforeman.org/contribute.html describes
+setting up a small test instance.
+
+You can enable the plugin right away, and see what it's default
+behaviour is, by editing foreman Gemfile.local.rb file (or creating this
+file under the folder bundler.d) and adding the following line
+
+.Gemfile.local.rb
+[source, ruby]
+----
+gem 'my_plugin', :path => 'path_to/my_plugin'
+----
+Install the 'preface' bundle by running
+
+[source]
+----
+bundle install
+----
+Restart (or start if it wasn't up) foreman (type 'rails server') and the
+new foreman plugin should be listed in the about page plugin tab. If it
+isn't, check your gem name and the symbol you passed to
+Foreman::Plugin.register match. Watch out for hyphens - e.g. gem
+'foreman-tasks' would need to be registered as Foreman::Plugin.register
+:"foreman-tasks".
+
+Note that Debian or other "production" installations need to be
+restarted after code changes, as they won't reload on the fly.
+
+[[rpm-installations]]
+=== RPM installations
+
+RPM installations use bundler_ext and are unable to load plugins from a
+path, they need the plugin to be built as a .gem file, installed and
+then reloaded. Development setups as described above are much better.
+
+In the plugin directory, run `gem build my_plugin.gemspec` which will
+build a file such as my_plugin-0.0.1.gem. Copy to the Foreman server and
+run
+`scl enable ruby193 "gem install --ignore-dependencies /tmp/my_plugin-0.0.1.gem"`
+
+Add to /usr/share/foreman/bundler.d/Gemfile.local.rb:
+[source, ruby]
+----
+gem 'my_plugin'
+----
+
+Then restart httpd to load it.
+
+[[initial-edits]]
+=== Initial edits
+
+First edit the my_plugin.gemspec file, you can specify here the name,
+authors, description homepage and version of your plugin, by simply
+replacing the appropriate strings with your content.
+
+Next is editing the lib/my_plugin/engine.rb file, add the following code
```

Comment:
The register block already exists in the foreman_plugin_template, so no need to add it.